File: 2Arrays and Strings/Code/Cpp/2-2IsSubsequence.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool isSubsequence(string s, string t) {
        int i = 0, j = 0;
        while (i < s.size() && j < t.size()) {
            if (s[i] == t[j]) {
                i++;
            }
            j++;
        }
        
        return i == s.size();
    }
};

int main() {
    Solution solution;
    
    bool result = solution.isSubsequence("ace", "abcde");
    cout << (result ? "True" : "False") << endl; // True

    result = solution.isSubsequence("aec", "abcde");
    cout << (result ? "True" : "False") << endl; // False

    result = solution.isSubsequence("abc", "ahbgdc");
    cout << (result ? "True" : "False") << endl; // True

    result = solution.isSubsequence("axc", "ahbgdc");
    cout << (result ? "True" : "False") << endl; // False
    return 0;
}

// Time-complexity : O(n+m)
// Space complexity: O(1)

File: 2Arrays and Strings/Code/Java/2-2IsSubsequence.java

class Solution {
    public boolean isSubsequence(String s, String t) {
        int i = 0;
        int j = 0;
        
        while (i < s.length() && j < t.length()) {
            if (s.charAt(i) == t.charAt(j)) {
                i++;
            }
            j++;
        }
        return i == s.length();
    }
}

class Main {
    public static void main(String[] args) {
	Solution solution = new Solution();
      
	System.out.println(solution.isSubsequence("ace", "abcde")); // true

      	System.out.println(solution.isSubsequence("aec", "abcde")); // false

	System.out.println(solution.isSubsequence("abc", "ahbgdc")); // true
      
      	System.out.println(solution.isSubsequence("axc", "ahbgdc")); // false
    }
};

// Compilation steps:
// javac 2-2IsSubsequence.java
// java Main

// Time-complexity : O(n+m)
// Space complexity: O(1)
